### PR TITLE
[develop]Scheduler plugin virtualenv

### PR DESCRIPTION
**Pull request by yuleiwan**
Signed-off-by: Yulei Wang <yuleiwan@amazon.com>

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cookbooks/aws-parallelcluster-scheduler-plugin/recipes/install_python.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+# License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+#
+install_pyenv node['cluster']['python-version'] do
+  prefix node['cluster']['scheduler_plugin']['pyenv_root']
```
Comment (lukeseawalker) [inline cookbooks/aws-parallelcluster-scheduler-plugin/recipes/install_python.rb:19]: cannot we reuse the pyenv installed in `prefix node['cluster']['system_pyenv_root']`?

File: attributes/default.rb [hunk 2]
```diff
@@ -136,6 +136,15 @@
 default['cluster']['scheduler_plugin']['shared_dir'] = "#{node['cluster']['shared_dir']}/scheduler-plugin"
 default['cluster']['scheduler_plugin']['local_dir'] = "#{node['cluster']['base_dir']}/scheduler-plugin"
 default['cluster']['scheduler_plugin']['scheduler_plugin_substack_outputs_path'] = "#{node['cluster']['shared_dir']}/scheduler-plugin-substack-outputs.json"
+default['cluster']['scheduler_plugin']['virtualenv'] = 'scheduler_plugin_virtualenv'
+default['cluster']['scheduler_plugin']['pyenv_root'] = "#{node['cluster']['scheduler_plugin']['home']}/pyenv"
+default['cluster']['scheduler_plugin']['virtualenv_path'] = [
+  node['cluster']['scheduler_plugin']['pyenv_root'],
```
Comment (lukeseawalker) [inline attributes/default.rb:142]: cannot we reuse the pyenv installed in `prefix node['cluster']['system_pyenv_root']`?

File: cookbooks/aws-parallelcluster-scheduler-plugin/resources/set_user_privileges.rb [hunk 3]
```diff
@@ -28,7 +28,7 @@
     Chef::Log.info("Set sudo privileges for #{node['cluster']['scheduler_plugin']['user']}")
     # Ensure scheduler user has sudoers capability
     template '/etc/sudoers.d/99-parallelcluster-scheduler-plugin' do
-      source 'byos_user/99-parallelcluster-scheduler-plugin.erb'
+      source 'scheduler_plugin_user/99-parallelcluster-scheduler-plugin.erb'
```
Comment (lukeseawalker) [inline cookbooks/aws-parallelcluster-scheduler-plugin/resources/set_user_privileges.rb:31]: this is addressed in https://github.com/aws/aws-parallelcluster-cookbook/pull/1227

File: cookbooks/aws-parallelcluster-scheduler-plugin/resources/execute_event_handler.rb [hunk 4]
```diff
@@ -141,6 +144,9 @@ def build_static_env(target_cluster_config, target_launch_templates, target_inst
     # PCLUSTER_INSTANCE_TYPES_DATA
     env.merge!({ 'PCLUSTER_INSTANCE_TYPES_DATA' => target_instance_types_data })
 
+    # PCLUSTER_PYTHON_VIRTUALENV
+    env.merge!({ 'PCLUSTER_PYTHON_VIRTUALENV' => target_python_virtualenv })
```
Comment (lukeseawalker) [inline cookbooks/aws-parallelcluster-scheduler-plugin/resources/execute_event_handler.rb:148]: how is the virtual env actually activated?

File: cookbooks/aws-parallelcluster-scheduler-plugin/recipes/init.rb [hunk 5]
```diff
@@ -18,6 +18,14 @@
 # create system user
 include_recipe "aws-parallelcluster-scheduler-plugin::init_user"
 
+# Add scheduler plugin virtualenv to default path
+template "/etc/profile.d/scheduler_plugin_virtualenv.sh" do
```
Comment (lukeseawalker) [inline cookbooks/aws-parallelcluster-scheduler-plugin/recipes/init.rb]: this is going to be set for all the users in this way.
Instead, this must be set only for the scheduler plugin user, let's find a way to set it only in the scheduler plugin user profile

**Review by demartinofra — APPROVED** [review]

**Pull request merged**